src/components/application_manager/rpc_plugins/sdl_rpc_plugin/src/commands/mobile/get_file_request.cc

```diff
@@ -0,0 +1,309 @@
+ LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "sdl_rpc_plugin/commands/mobile/get_file_request.h"
+#include "application_manager/application_impl.h"
+#include "application_manager/rpc_service.h"
+#include "interfaces/MOBILE_API.h"
+#include "application_manager/message_helper.h"
+#include "application_manager/app_service_manager.h"
+#include "utils/file_system.h"
+#include <boost/crc.hpp>
+
+namespace {
+/**
+* Calculates CRC32 checksum
+* @param binary_data - input data for which CRC32 should be calculated
+* @return calculated CRC32 checksum
+*/
+uint32_t GetCrc32CheckSum(const std::vector<uint8_t>& binary_data) {
+  const std::size_t file_size = binary_data.size();
+  boost::crc_32_type result;
+  result.process_bytes(&binary_data[0], file_size);
+  return result.checksum();
+}
+
+}  // namespace
+
+namespace sdl_rpc_plugin {
+using namespace application_manager;
+namespace commands {
+
+GetFileRequest::GetFileRequest(
+    const application_manager::commands::MessageSharedPtr& message,
+    ApplicationManager& application_manager,
+    app_mngr::rpc_service::RPCService& rpc_service,
+    app_mngr::HMICapabilities& hmi_capabilities,
+    policy::PolicyHandlerInterface& policy_handler)
+    : CommandRequestImpl(message,
+                         application_manager,
+                         rpc_service,
+                         hmi_capabilities,
+                         policy_handler)
+    , file_name_("")
+    , file_type_(mobile_apis::FileType::INVALID_ENUM)
+    , length_(0)
+    , offset_(0) {}
+
+GetFileRequest::~GetFileRequest() {}
+
+bool GetFileRequest::GetFilePath(std::string& file_path, bool& forward_to_hmi) {
+  std::string path = application_manager_.get_settings().app_storage_folder();
+  auto connect_key = connection_key();
+  forward_to_hmi = false;
+
+  if ((*message_)[strings::msg_params].keyExists(strings::app_service_id)) {
+    std::string service_id =
+        (*message_)[strings::msg_params][strings::app_service_id].asString();
+    LOG4CXX_DEBUG(logger_,
+                  "Finding storage directory for service id: " << service_id);
+
+    AppService app_service_info;
+    if (application_manager_.GetAppServiceManager().GetAppServiceInfo(
+            service_id, app_service_info)) {
+      if (app_service_info.mobile_service) {
+        connect_key = app_service_info.connection_key;
+      } else {
+        forward_to_hmi = true;
+        return true;
+      }
+    } else {
+      return false;
+    }
+  } else {
+    LOG4CXX_DEBUG(logger_, "Using current storage directory");
```

Comment:
I get the feeling that using the current app's directory would be unexpected behavior if `appServiceID` was omitted. That said, I don't know where the correct place to search would be, so perhaps we should leave a note of this in the API.

Reply:
I was thinking that GetFile could be used as a general RPC outside of app services(when a mobile app wants to get a file it uploaded and which was then modified by core perhaps). Since the service_id is an optional parameter not sure how else `GetFile` would be handled if service_id was not specified

Reply:
Yeah, I'm just thinking a note in the `serviceID` description within the API would be good